rust/src/operations/mod.rs

```diff
@@ -1,329 +1,121 @@
-        }
-    }
-}
+#[cfg(feature = "datafusion-ext")]
+mod load;
+#[cfg(feature = "datafusion-ext")]
+pub mod write;
+// TODO the writer module does not actually depend on datafusion,
+// eventually we should consolidate with the record batch writer
+#[cfg(feature = "datafusion-ext")]
+mod writer;
 
-fn to_datafusion_err(e: impl std::error::Error) -> DataFusionError {
-    DataFusionError::Plan(e.to_string())
-}
+/// Maximum supported writer version
+pub const MAX_SUPPORTED_WRITER_VERSION: i32 = 1;
+/// Maximum supported reader version
+pub const MAX_SUPPORTED_READER_VERSION: i32 = 1;
 
 /// High level interface for executing commands against a DeltaTable
-pub struct DeltaCommands {
-    table: DeltaTable,
-}
-
-impl DeltaCommands {
-    /// load table from uri
-    pub async fn try_from_uri<T>(uri: T) -> DeltaCommandResult<Self>
-    where
-        T: Into<String>,
-    {
-        let table_uri: String = uri.into();
-        let table = if let Ok(tbl) = open_table(&table_uri).await {
-            Ok(tbl)
-        } else {
-            DeltaTableBuilder::from_uri(table_uri).build()
-        }?;
-        Ok(Self { table })
-    }
-
-    /// Get a reference to the underlying table
-    pub fn table(&self) -> &DeltaTable {
-        &self.table
+pub struct DeltaOps(pub DeltaTable);
+
+impl DeltaOps {
+    /// Create a new [`DeltaOps`] instance, operating on [`DeltaTable`] at given uri.
+    ///
+    /// ```
+    /// use deltalake::DeltaOps;
+    ///
+    /// async {
+    ///     let ops = DeltaOps::try_from_uri("memory://").await.unwrap();
+    /// };
+    /// ```
+    pub async fn try_from_uri(uri: impl AsRef<str>) -> DeltaResult<Self> {
+        let mut table = DeltaTableBuilder::from_uri(uri).build()?;
+        // We allow for uninitialized locations, since we may want to create the table
+        match table.load().await {
+            Ok(_) => Ok(table.into()),
+            Err(DeltaTableError::NotATable(_)) => Ok(table.into()),
+            Err(err) => Err(err),
+        }
     }
 
-    async fn execute(
-        &mut self,
-        operation: DeltaOperation,
-        plan: Arc<dyn ExecutionPlan>,
-    ) -> DeltaCommandResult<()> {
-        let transaction = Arc::new(DeltaTransactionPlan::new(
-            self.table.table_uri(),
-            self.table.version(),
-            plan,
-            operation,
-            None,
-        ));
-
-        let session_ctx = SessionContext::new();
-        let task_ctx = session_ctx.task_ctx();
-        let _ = collect(transaction, task_ctx).await?;
-        self.table.update().await?;
-
-        Ok(())
+    /// Create a new [`DeltaOps`] instance, backed by an un-initialized in memory table
+    ///
+    /// Using this will not persist any changes beyond the lifetime of the table object.
+    /// THe main purpose of in-memory tables is for use in testing.
+    ///
+    /// ```
+    /// use deltalake::DeltaOps;
+    ///
+    /// let ops = DeltaOps::new_in_memory();
+    /// ```
+    #[must_use]
```

Comment:
This is new to me. Why `#[must_use]`?

Reply:
I also only fairly recently learned about must use. The idea is, if you do not consume the result of this call, clippy (or even the compiler) will complain. which to me makes sense, since if not using the return it literally does nothing and should be removed.

I think futures for instance may also be must_use since if not consumed they also so nothing ...